packages/language-chooser/src/language-chooser/index.tsx

```diff
@@ -0,0 +1,338 @@
+/**
+ * External dependencies
+ */
+import type { KeyboardEvent } from 'react';
+
+/**
+ * WordPress dependencies
+ */
+import { useEffect, useState } from '@wordpress/element';
+import { __, sprintf } from '@wordpress/i18n';
+import { speak } from '@wordpress/a11y';
+import { Notice } from '@wordpress/components';
+import { useInstanceId } from '@wordpress/compose';
+
+/**
+ * Internal dependencies
+ */
+import ActiveLocales from './active-locales';
+import InactiveLocales from './inactive-locales';
+import type { Language } from './types';
+import { reorder } from './utils';
+
+function MissingTranslationsNotice() {
+	return (
+		<Notice status="warning" isDismissible={ false }>
+			{ __(
+				'Some of the languages are not installed. Re-save changes to download translations.'
+			) }
+		</Notice>
+	);
+}
+
+interface HiddenFormFieldProps {
+	preferredLanguages: Language[];
+	inputName: string;
+}
+
+function HiddenFormField( {
+	preferredLanguages,
+	inputName,
+}: HiddenFormFieldProps ) {
+	const value = preferredLanguages
+		.filter( ( language ) => Boolean( language ) )
+		.map( ( { locale } ) => locale )
+		.join( ',' );
+
+	return <input type="hidden" name={ inputName } value={ value } />;
+}
+
+interface LanguageChooserProps {
+	allLanguages: Language[];
+	preferredLanguages: Language[];
```

Comment:
The reason why I mentioned supporting both controlled and uncontrolled usage is because, if not done from the start, it's difficult to backtrack at a later point ([example](https://github.com/WordPress/gutenberg/issues/57004)).

You could take inspiration from [the `useControlledValue` utility hook](https://github.com/WordPress/gutenberg/blob/8dee3e6764da100b20ad9abd41af5df086b5f42f/packages/components/src/utils/hooks/use-controlled-value.ts). And if you opted for using `Ariakit` components, they also offer this functionality [out of the box](https://ariakit.org/guide/component-stores#controlled-state)